### PR TITLE
Document amppackager's handlers and how to monitor them 

### LINKED ISSUE #436: Document amppackager's handlers and how to monitor them

### PR DESCRIPTION
**Pull request by MichaelRybak**
Resolves #436.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -264,6 +264,18 @@ against the [AMP Cache requirement](docs/cache_requirements.md) for a minimum
 `max-age` of `345600` (4 days). For SXGs shorter than that, the Google AMP Cache
 will treat them as if unsigned (by showing an AMP Viewer).
 
+#### Monitoring `amppackager` in production
+
+Once you've run the `amppackager` server in production, you may want to monitor its
+health and performance. You may also monitor the performance of the underlying requests to the AMP document server. You can monitor both servers by curling `/healthz` and `/metrics` endpoints provided by `amppackager`. A few examples of questions you can answer:
+* Is `amppackager` up and running?
```
Comment (banaag) [inline README.md:271]: please follow style guide: https://g3doc.corp.google.com/corp/g3doc/docs/reference/style.md?cl=head#lists

File: README.md [hunk 2]
```diff
@@ -264,6 +264,18 @@ against the [AMP Cache requirement](docs/cache_requirements.md) for a minimum
...
+* How many requests has it processed since it's been up?
+* What was the 0.9 percentile latency of handling those request?
+* How many of those requests have triggered a gateway request to the AMP document server?
+* For those gateway requests, what was the 0.9 percentile latency of the AMP document server?
+
+These are just a few examples. To learn what are all the things you can monitor, and how to do it, check the [monitoring manual](monitoring.md).
+
 ## Local Transformer
 
 The local transformer is a library within the AMP Packager that transforms AMP
```
Comment (banaag) [inline README.md:281]: Add a period at the end.

File: monitoring.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+Once you've run `amppackager` server in production, you may want to monitor its
+health and performance, as well as the performance of the underlying requests to the AMP document server. 
+
+The sections below talk about the things you can monitor, and how to do it.
+
+All the command line examples are for Linux OS.
+
+## Monitoring health
+
+To check `amppackager`'s health curl its `/healthz` endpoint:
```
Comment (banaag) [inline monitoring.md:12]: health -> health, also should curl be enclosed in back ticks?

File: monitoring.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+## Monitoring performance
+
+You can take a step further and check a few performance metrics, both for requests
+handled by `amppackager`, and for the underlying gateway requests that 
+`amppackager` sends to the AMP document server. Get all the metrics by curling the `/metrics` endpoint. 
+
+## Example: monitoring total requests count
+
+The example command below fetches all the available metrics. It then greps the report for `total_requests_by_code_and_url` metric. This metric counts the HTTP requests the `amppackager` server has processed since it's been up. 
+
```
Comment (banaag) [inline monitoring.md:29]: Enclosed the block in back ticks?

File: monitoring.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    # HELP total_requests_by_code_and_url Total number of requests by HTTP code and URL.
+    # TYPE total_requests_by_code_and_url counter
+    total_requests_by_code_and_url{code="200",handler="healthz"} 3
+    total_requests_by_code_and_url{code="200",handler="validityMap"} 5
+    total_requests_by_code_and_url{code="200",handler="signer"} 6
+    total_requests_by_code_and_url{code="502",handler="signer"} 4
+    total_requests_by_code_and_url{code="404",handler="handler_not_assigned"} 1
+ 
+ The example stats above are broken down by response HTTP code, and by the internal amppackager's module (handler) that has handled the request. The stats report 3 requests to the `healthz` handler that got a 200 response (OK), 4 requests to the `signer` handler that got a 502 response (Bad Gateway) etc.
```
Comment (banaag) [inline monitoring.md:40]: > 80 chars
go/g3doc-style#80

File: monitoring.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+## `amppackager`'s handlers 
+
+The table below lists `amppackager`'s handlers accounted for by the metrics:
+
+| Handler | Responsibility |
+|-|-|
+| signer | Handles `/priv/doc` requests. Fetches the AMP document, signs it and returns it. |
+| certCache | Handles `/amppkg/cert` requests. Returns your Signed Exchange certificate. |
+| validityMap | Handles `/amppkg/validity` requests. Returns the [validity data](https://tools.ietf.org/html/draft-yasskin-httpbis-origin-signed-exchanges-impl-00#section-3.6) referred by `amppackager`'s signatures. |
```
Comment (banaag) [inline monitoring.md:50]: > 80 chars

File: monitoring.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+| healthz | Handles `/healthz` requests. Checks if `amppackager` is running and has a valid, fresh certificate. |
+| metrics | Handles `/metrics` requests. Reports performance metrics for `amppackager` and for the underlying gateway requests to the AMP document server. |
+
+## Metrics labels: breakdown by handler and response code
+
+For some metrics like `total_requests_by_code_and_url` the stats in the `/metrics` response are grouped into buckets by two dimensions: the handler name, and the HTTP response code. E.g. note the buckets in the the example above:
+
```
Comment (banaag) [inline monitoring.md:57]: backtick the block?

File: monitoring.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    total_requests_by_code_and_url{code="200",handler="healthz"} 3
+    total_requests_by_code_and_url{code="200",handler="signer"} 6
+    total_requests_by_code_and_url{code="502",handler="signer"} 4
+
+Invalid requests that were not routed by `amppackager` to any handler are assigned a special label `handler_not_assigned`:
+
+    total_requests_by_code_and_url{code="404",handler="handler_not_assigned"} 1
+
+ Some metrics only make sense for a particular handler. E.g. `gateway_request_latencies_in_seconds` and other metrics related to gateway requests are only related to `signer` handler's operation. Such metrics are only broken down into buckets by the response code, not by the handler.
```
Comment (banaag) [inline monitoring.md:66]: > 80 chars

File: monitoring.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+ Some metrics only make sense for a particular handler. E.g. `gateway_request_latencies_in_seconds` and other metrics related to gateway requests are only related to `signer` handler's operation. Such metrics are only broken down into buckets by the response code, not by the handler. 
+ 
+ __Labels__ are key-value properties of buckets that indicate the specific values of the breakdown dimensions, e.g. `code="200"` or `handler="healthz"`.
+
+## Metrics types: counters and summaries
+
+The two types of metrics are *counters* and *summaries*. 
+
+For some metrics like `total_requests_by_code_and_url`, each request increases the total by 1, so the metric is a *counter* that has no historical data, just the accumulated total.
```
Comment (banaag) [inline monitoring.md:74]: > 80 chars, same below.

File: monitoring.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+|--|--|--|--|--|
+| total_requests_by_code_and_url | Counter | Total number of requests handled by `amppackager` since it's been up. | Yes | Yes |
+| request_latencies_in_seconds | Summary | `amppackager`'s handlers latencies in seconds, measured from the moment the handler starts processing the request, to the moment the response is returned. | Yes | Yes |
+| total_gateway_requests_by_code | Counter | Total number of underlying requests sent by `signer` handler to the AMP document server. | Yes | No, specific to `signer` handler. |
+| gateway_request_latencies_in_seconds | Summary | Latencies (in seconds) of gateway requests to the AMP document server. | Yes | No, specific to `signer` handler. |
+
+## Understanding percentiles reported by Summaries
+
+A [percentile](https://en.wikipedia.org/wiki/Percentile) indicates the value below which a given percentage of observations in a group of observations falls. E.g. if in a room of 100 people, 80% are shorter than you, then you are the 80% percentile. The 50% percentile is also known as [median](https://en.wikipedia.org/wiki/Median).
```
Comment (banaag) [inline monitoring.md:91]: > 80 chars, same below.

File: monitoring.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+For summary metrics like `request_latencies_in_seconds`, the `/metrics` endpoint provides three percentiles: 0.5, 0.9, 0.99. 
+
+To get an idea of how long it __usually__ takes `amppackager` to handle a request, look at the respective 0.5 percentile. To check the rare, __worst case scenarios__, look at 0.9 and 0.99 percentiles.
+
+Consider the following example. Let's say you're interested in the stats for the `request_latencies_in_seconds` metric, specifically for requests that got an OK response (200) from the `signer` handler:
+
+    $ curl https://127.0.0.1:8080/metrics | grep request_latencies_in_seconds | grep signer | grep code=\"200\"
```
Comment (banaag) [inline monitoring.md:99]: backtick enclose the block

File: monitoring.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+In the example above all the 10000 requests were handled in 661 seconds, which means the mean (average) latency was 0.0661s. This value is ~3 times larger than the median. So which one more accurately represents the "typical" scenario? Why not look at mean instead of looking at percentiles?
+
+Median (0.5 percentile) is [more stable against outliers](https://en.wikipedia.org/wiki/Median) than mean, and therefore gives a better understanding of the typical response time. At the same time the 0.9 and 0.99 percentiles give you a good idea about the large outliers, i.e. abnormally slow response times.
+
+## More examples
+
+Check all the stats (note - this command produces numerous metrics that are self-documented; the key ones are also explained in the table above):
+
+    $ curl https://127.0.0.1:8080/metrics
```
Comment (banaag) [inline monitoring.md:125]: backtick all code blocks

File: monitoring.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    $ curl https://127.0.0.1:8080/metrics | grep gateway_request_latencies_in_seconds | grep code=\"200\"
+
+Check the 0.9 percentile latency for the `request_latencies_in_seconds` metric, for requests that got an OK response (200) from the `signer` handler:
+
+    $ curl https://127.0.0.1:8080/metrics | grep request_latencies_in_seconds | grep signer | grep code=\"200\" | grep quantile=\"0.9\"
+
+
+## Performance metrics lifetime
+
+The metrics provided by the `/metrics` endpoint are reset upon the execution of the `amppackager` server binary. Every request to `/metrics` is served with the stats accumulated since the server's been up, up to the time of the request, but not including the request itself.
```
Comment (banaag) [inline monitoring.md:142]: > 80 chars

File: monitoring.md (hunk 5 above)
Comment (MichaelRybak) [reply monitoring.md:40]: Thanks for the comments, all done.

File: README.md [hunk 14]
```diff
@@ -199,8 +199,9 @@ that:
 
 It is possible to test an otherwise fully production configuration without
 obtaining a certificate with the `CanSignHttpExchanges` extension. `amppkg`
-still needs to perform OCSP verification, so the Issuer CA must be valid (i.e. no
-self-signed certificates). e.g. You can use a certificate from [Let's Encrypt](https://letsencrypt.org/).
+still needs to perform OCSP verification, so the Issuer CA must be valid (i.e.
```
Comment (MichaelRybak) [inline README.md:202]: Enforced 80 chars limit.

File: README.md (hunk 1 above)
Comment (MichaelRybak) [reply README.md]: Done.

File: monitoring.md (hunk 3 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 4 above)
Comment (MichaelRybak) [reply monitoring.md:32]: Done.

File: monitoring.md (hunk 5 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 6 above)
Comment (MichaelRybak) [reply monitoring.md:58]: Done.

File: monitoring.md (hunk 7 above)
Comment (MichaelRybak) [reply monitoring.md:68]: Done.

File: monitoring.md (hunk 8 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 9 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 10 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 11 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 12 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 13 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+The sections below talk about the things you can monitor, and how to do it.
+
+All the command line examples are for Linux OS.
+
+## Monitoring health
+
+To check `amppackager`'s health, `curl` its `/healthz` endpoint:
+
+```
```
Comment (twifkak) [inline monitoring.md:15]: Might try appending `shellsession` to this line (or one of its [aliases](https://github.com/github/linguist/blob/2a6d952df870807034c3830738f95ad33222aca0/lib/linguist/languages.yml#L5112-L5113)) to get [syntax highlighting](https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks#syntax-highlighting). (Ditto the rest.)

File: monitoring.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+If the server is up and has a fresh, valid certificate, it will respond with
+`ok`. If not, it will provide an error message.
+
+## Monitoring performance
+
+You can take a step further and check a few performance metrics, both for
+requests handled by `amppackager`, and for the underlying gateway requests that
+`amppackager` sends to the AMP document server. Get all the metrics by `curl`ing
+the `/metrics` endpoint.
```
Comment (twifkak) [inline monitoring.md:27]: You should mention somewhere in this doc that this is a Prometheus endpoint, and that the preferred way of querying these metrics is to set up all amppkg replicas as targets and [query](https://prometheus.io/docs/prometheus/latest/querying/basics/) the prom server or [draw graphs](https://prometheus.io/docs/visualization/grafana/). (Or whatever the correct lingo is.)

File: monitoring.md (hunk 15 above)
Comment (MichaelRybak) [reply monitoring.md]: Done.

File: monitoring.md (hunk 16 above)
Comment (MichaelRybak) [reply monitoring.md:47]: Done.

**Pull request merged**